### PR TITLE
1900 nlp tf idf drift check PR draft

### PR DESCRIPTION
**Pull request by TheSolY**
basic functionality

### CODE REVIEW

**Comment by TheSolY** [comment]
image

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+from sklearn.feature_extraction.text import TfidfVectorizer
+from nltk.stem.porter import PorterStemmer
+import numpy as np
+
+__all__ = ['KeywordFrequencyDrift']
+
+
+class KeywordFrequencyDrift(TrainTestCheck):
+    """
+    Computes the keywords' frequencies drift between the train and the test datasets.
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:30]: We'd probably want to expand the description here a bit

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    Parameters
+    ----------
+    top_n_to_show: int, default: 20
+        How many words will be displayed in the graph.
+    top_n_method: Union[str, List[str]], default: 'top_diff'
+        Decides which method will be used to select the top n words to show. Possible values are:
+        "top_diff" for the top difference between train and test, "top_freq" for the top absolute frequencies, or a list
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:38]: Can format that like line 93 in `train_test_prediction_drift.py`, with bullet points per option

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        of keywords in which case the words on the list are shown and the "top_n_to_show" parameter is ignored.
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:39]: IMO we should have two independent arguments - one controlling top_n_method, and one enabling to add pre-selected keywords. My thinking here is that the user may want to monitor some keywords, but it still interested in watching other words that have changed dramatically

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        super().__init__(**kwargs)
+        self.top_n_to_show = top_n_to_show
+        self.top_n_method = top_n_method
+
+        if drift_method == 'PSI':
+            self.drift_method = psi
+        elif drift_method == 'cramer_v':
+            self.drift_method = cramers_v
+        else:
+            raise DeepchecksValueError('drift_method must be one of: PSI, cramer_v')
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:59]:
```suggestion
            raise DeepchecksValueError(f'drift_method must be one of: PSI, cramer_v, found {drift_method}')
```

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        self.stem_func = PorterStemmer().stem
+        self.token_pattern = r'[a-z]{2,}'
+
+        self.top_n_words = None
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:63]: Think these two are not used anywhere

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+from deepchecks.core import DatasetKind
+from deepchecks.utils.strings import format_number
+from deepchecks.core.errors import DeepchecksValueError
+from deepchecks.utils.distribution.drift import word_counts_drift_plot
+from deepchecks.nlp.base_checks import TrainTestCheck
+from deepchecks.utils.distribution.drift import cramers_v, psi
+from typing import Union, List, Any
+from deepchecks.nlp.context import Context
+from sklearn.feature_extraction.text import TfidfVectorizer
+from nltk.stem.porter import PorterStemmer
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:22]: add nltk to `nlp-requirements.txt` (if we decide to use it eventually)

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        all_data = list(train_dataset.text) + list(test_dataset.text)
+
+        vectorizer = TfidfVectorizer(input='content', strip_accents='ascii', preprocessor=self.stem_func,
+                                     token_pattern=self.token_pattern)
+        vectorizer.fit(all_data)
+        train_freqs = vectorizer.transform(train_dataset.text)
+        max_train_freqs = np.array(train_freqs.max(axis=0).todense()).reshape(-1)
+
+        test_freqs = vectorizer.transform(test_dataset.text)
+        max_test_freqs = np.array(test_freqs.max(axis=0).todense()).reshape(-1)
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:79]: Why look at the max tfidf value over all samples, and not the mean one?

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        top_n_diffs = np.take(word_freq_diff, top_n_idxs)
+
+        if context.with_display:
+            train_to_show = max_train_freqs[top_n_idxs]
+            test_to_show = max_test_freqs[top_n_idxs]
+            display = word_counts_drift_plot(train_to_show, test_to_show, top_n_words)
+        else:
+            display = None
+
+        result = {'drift_score': drift_score, 'top_n_diffs': dict(zip(top_n_words, top_n_diffs))}
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:108]: Feel like it would be preferable to get the differences (not the np.abs of differences) here, if the user wants to use the results somehow.

File: deepchecks/nlp/context.py [hunk 9]
```diff
@@ -289,8 +289,8 @@ def __init__(
             if test_dataset.has_label() and train_dataset.has_label() and not \
                     TextData.datasets_share_task_type(train_dataset, test_dataset):
                 raise DatasetValidationError('train_dataset and test_dataset must share the same label and task type')
-            if train_dataset.name == test_dataset.name:
-                raise DatasetValidationError('train_dataset and test_dataset must have different names')
+            # if train_dataset.name == test_dataset.name:
```
Comment (noamzbr) [inline deepchecks/nlp/context.py:292]: Agree, you can just delete that instead of commenting out.

File: deepchecks/utils/distribution/drift.py [hunk 10]
```diff
@@ -118,7 +125,8 @@ def cramers_v(dist1: Union[np.ndarray, pd.Series], dist2: Union[np.ndarray, pd.S
 
 
 def psi(dist1: Union[np.ndarray, pd.Series], dist2: Union[np.ndarray, pd.Series],
-        min_category_size_ratio: float = 0, max_num_categories: int = None, sort_by: str = 'dist1') -> float:
+        min_category_size_ratio: float = 0, max_num_categories: int = None, sort_by: str = 'dist1',
+        from_freqs: bool = True) -> float:
```
Comment (noamzbr) [inline deepchecks/utils/distribution/drift.py:129]: Add docstring for `from_freqs` here as well

File: deepchecks/utils/distribution/drift.py [hunk 11]
```diff
@@ -89,14 +91,19 @@ def cramers_v(dist1: Union[np.ndarray, pd.Series], dist2: Union[np.ndarray, pd.S
         - 'dist2': Sort by the largest dist2 categories.
         - 'difference': Sort by the largest difference between categories.
         > Note that this parameter has no effect if max_num_categories = None or there are not enough unique categories.
+    from_freqs: bool, default: False
```
Comment (noamzbr) [inline deepchecks/utils/distribution/drift.py:94]: IMO from_freq is not the right name. For me it implies that the values are already normalized 0-1 or something. Actually if I understood correctly you're saying it already comes as counts rather than raw values, so maybe `from_value_counts` or something along these lines?

File: tests/nlp/nlp_conftest.py [hunk 12]
```diff
@@ -36,3 +39,28 @@ def text_multilabel_classification_dataset_mock():
     return TextData(['I think therefore I am', 'I am therefore I think', 'I am'],
                     [[0, 0, 1], [1, 1, 0], [0, 1, 0]],
                     task_type='text_classification')
+
+
+@pytest.fixture(scope='session')
+def movie_reviews_data():
```
Comment (noamzbr) [inline tests/nlp/nlp_conftest.py:45]: How heavy is that dataset? Maybe worth considering using a subset of it

File: deepchecks/utils/distribution/plot.py [hunk 13]
```diff
@@ -362,3 +362,27 @@ def sort_int_and_strings(a, b):
     )
 
     return traces, yaxis_layout
+
+def word_counts_bar_traces(
```
Comment (noamzbr) [inline deepchecks/utils/distribution/plot.py:366]: What's the difference between this and the regular categorical drift plot?

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        else:
+            raise DeepchecksValueError('top_n_method must be one of: top_diff, top_freq or a list of keywords')
+
+        top_n_words = np.take(np.array(vocab), top_n_idxs)
+        top_n_diffs = np.take(word_freq_diff, top_n_idxs)
+
+        if context.with_display:
+            train_to_show = max_train_freqs[top_n_idxs]
+            test_to_show = max_test_freqs[top_n_idxs]
+            display = word_counts_drift_plot(train_to_show, test_to_show, top_n_words)
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:104]: I think we should add here some text explaining what the checks does, same as we do in https://docs.deepchecks.com/stable/checks_gallery/tabular/train_test_validation/plot_train_test_feature_drift.html and other complicated checks. Explain what is a drift score (can copy the paragraph from the regular drift check), that we're showing the difference in TF-IDF and what it means.

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+        Decides which method will be used to select the top n words to show. Possible values are:
+        "top_diff" for the top difference between train and test, "top_freq" for the top absolute frequencies, or a list
+        of keywords in which case the words on the list are shown and the "top_n_to_show" parameter is ignored.
+    drift_method: str, default: "cramer_v"
+        Decides which method will be used for drift calculation. Possible values are:
+        "cramer_v" for Cramer's V, "PSI" for Population Stability Index (PSI).
+    """
+
+    def __init__(self,
+                 top_n_to_show: int = 20,
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:46]: Think we also need n_samples and random_state here, no need to run on all the data
Unless this will become obsolete by https://github.com/deepchecks/deepchecks/issues/1942 and then please ignore my comment

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        vectorizer.fit(all_data)
+        train_freqs = vectorizer.transform(train_dataset.text)
+        max_train_freqs = np.array(train_freqs.max(axis=0).todense()).reshape(-1)
+
+        test_freqs = vectorizer.transform(test_dataset.text)
+        max_test_freqs = np.array(test_freqs.max(axis=0).todense()).reshape(-1)
+        word_freq_diff = np.abs(max_train_freqs - max_test_freqs)
+
+        max_train_counts = max_train_freqs * test_dataset.n_samples
+        max_test_counts = max_test_freqs * test_dataset.n_samples
```
Comment (noamzbr) [inline deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:83]: This promises a scale of 0 to 1?

**Review by noamzbr — COMMENTED** [review]
Looks good! Main issue for now is making sure the stemmer runs fast, the rest of the comments can wait

File: deepchecks/utils/distribution/plot.py (hunk 13 above)
Comment (TheSolY) [reply deepchecks/utils/distribution/plot.py:366]: @noamzbr the regular categorical accepts a categorical series and calculates the counts and the train-test alignment as well, here the data is already aligned and in frequencies

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py (hunk 16 above)
Comment (TheSolY) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:91]: The opposite - this makes it in total counts instead of fraction, the format that both Cramer's V and PSI expect.

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py (hunk 1 above)
Comment (TheSolY) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:34]: @noamzbr agreed, for the final version

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py (hunk 3 above)
Comment (TheSolY) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:43]: In that case, how would you suggest to combine between the keywords and the top n?

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py (hunk 5 above)
Comment (TheSolY) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:67]: I'll remove them after setting the final implementation

File: deepchecks/nlp/context.py (hunk 9 above)
Comment (noamzbr) [reply deepchecks/nlp/context.py:292]: On a second thought, I rely on the fact that dataset have different names to retrieve the appropriate prediction indexes from the prediction and proba vectors, so I do need the names not to be the same

Comment (TheSolY) [reply deepchecks/nlp/context.py:292]: @noamzbr the name is None for both before the context assigns "Train" or "test" by default so they're the same at the point of the validation if a name is not specified.

File: deepchecks/utils/distribution/drift.py (hunk 11 above)
Comment (TheSolY) [reply deepchecks/utils/distribution/drift.py:95]: When doing statistic tests the commonly used term is frequencies https://en.wikipedia.org/wiki/Frequency_(statistics)

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py (hunk 8 above)
Comment (TheSolY) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:110]: we need this for the condition

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py (hunk 3 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py]: I suggest that the keywords will be "on top" of the words that have top diff / top values in TFIDF. This suggestion has the drawback of the possibility of having to many words for the display (if the user adds 10 keywords, they'll have to reduce the n_to_show to get a normal display) . Open to other suggestions.

File: deepchecks/utils/distribution/plot.py (hunk 13 above)
Comment (noamzbr) [reply deepchecks/utils/distribution/plot.py:367]: So maybe let's put this plot in the drift.py file? Not a must, only if you think we may use this elsewhere

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py (hunk 16 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:94]: Oh right, and the plot is of the version before multiplying by n_samples?

Comment (TheSolY) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:94]: correct

File: deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py (hunk 8 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/train_test_validation/keyword_frequency_drift.py:125]: condition can compute the abs internally

File: deepchecks/nlp/text_data.py [hunk 17]
```diff
@@ -87,17 +87,22 @@ def __init__(
...
             if label is not None:
-                raise DeepchecksValueError('task_type must be set when label is provided')
-            self._task_type = None
+                if isinstance(label, t.Sequence):
+                    if pd.notnull(label).any():
+                        raise DeepchecksValueError('task_type must be set when label is provided')
+                else:
+                    raise DeepchecksValueError('task_type must be set when label is provided')
+
+            self._task_type = TaskType.OTHER
```
Comment (noamzbr) [inline deepchecks/nlp/text_data.py:98]: Did you add OTHER to the enum? Can't find it here

File: tests/nlp/nlp_conftest.py [hunk 18]
```diff
@@ -36,3 +39,31 @@ def text_multilabel_classification_dataset_mock():
...
+    """Dataset of single sentence samples."""
+    sentences = [' '.join(x) for x in movie_reviews.sents()]
+    random.seed(42)
+    train_data = TextData(random.choices(sentences, k=10000))
```
Comment (noamzbr) [inline tests/nlp/nlp_conftest.py:49]: Why use 10K here and not 1K like the two other fixtures?

File: deepchecks/nlp/text_data.py (hunk 17 above)
Comment (TheSolY) [reply deepchecks/nlp/text_data.py:98]: It was already there, I didn't need to add

File: tests/nlp/nlp_conftest.py (hunk 18 above)
Comment (TheSolY) [reply tests/nlp/nlp_conftest.py:49]: @noamzbr because there's not enough overlap in the random split to show that it's more similar than split by label with less then 10k

**Review by noamzbr — APPROVED** [review]
